### PR TITLE
Upgrade to latest mimir-prometheus

### PR DESCRIPTION
**Pull request by aknuds1**
#### What this PR does
Upgrade to latest mimir-prometheus, see [PR](https://github.com/grafana/mimir-prometheus/pull/539) for its latest sync with upstream. I'm not aware of any behavioral changes that'll affect Mimir.

Please note that I'm replacing github.com/charleskorn/goautoneg  with a Grafana Labs fork github.com/grafana/goautoneg, based off of the same commit. I needed to make a fix to goautoneg so it would compile with the new revision of golang.org/x/exp. I think it's better going forward to use a Grafana Labs fork rather than a personal one?

#### Checklist

- [x] Tests updated
- [na] Documentation added
- [na] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/cardinality/request.go [hunk 1]
```diff
@@ -183,14 +183,18 @@ func extractSelector(values url.Values) (matchers []*labels.Matcher, err error)
 	}
 
 	// Ensure stable sorting (improves query results cache hit ratio).
-	slices.SortFunc(matchers, func(a, b *labels.Matcher) bool {
-		if a.Name != b.Name {
-			return a.Name < b.Name
+	slices.SortFunc(matchers, func(a, b *labels.Matcher) int {
+		nameCmp := strings.Compare(a.Name, b.Name)
```
Comment (pracucci) [inline pkg/cardinality/request.go:187]: Never use `strings.Compare()` please. See the doc:
https://cs.opensource.google/go/go/+/refs/tags/go1.21.2:src/strings/compare.go;l=13

Always convert string comparison into a `<`, `==`, `>`.

Comment (pracucci) [reply pkg/cardinality/request.go:187]: You can keep using `strings.Compare()` in tests. We don't care about performance there.

Comment (aknuds1) [reply pkg/cardinality/request.go:187]: Thanks for the heads up, I wasn't aware. Will fix.

File: pkg/frontend/querymiddleware/labels_query_cache.go [hunk 2]
```diff
@@ -154,22 +154,25 @@ func parseRequestMatchersParam(values url.Values, paramName string) ([][]*labels
...
 		for ; idx < len(a) && idx < len(b); idx++ {
 			if c := compareLabelMatchers(a[idx], b[idx]); c != 0 {
-				return c < 0
+				return c
 			}
 		}
 
 		// All label matchers are equal so far. Check which one has fewer matchers.
-		return idx < len(b)
+		if idx < len(b) {
```
[the remark this answers is not in the document]
Comment (aknuds1) [reply pkg/frontend/querymiddleware/labels_query_cache.go:172]: Looking into it.

File: vendor/github.com/prometheus/prometheus/tsdb/index/postings.go [hunk 3]
```diff
@@ -108,11 +109,14 @@ func (p *MemPostings) SortedKeys() []labels.Label {
 	}
 	p.mtx.RUnlock()
 
-	slices.SortFunc(keys, func(a, b labels.Label) bool {
-		if a.Name != b.Name {
-			return a.Name < b.Name
+	slices.SortFunc(keys, func(a, b labels.Label) int {
+		nameCompare := strings.Compare(a.Name, b.Name)
```
Comment (pracucci) [inline vendor/github.com/prometheus/prometheus/tsdb/index/postings.go:113]: Would be great to open an upstream PR to not use `strings.Compare()` here. This should be something like:

```
if a.Name == b.Name {
  // ... compare using < > here ...
}
```

File: pkg/cardinality/request.go (hunk 1 above)
Comment (aknuds1) [reply pkg/cardinality/request.go:187]: @pracucci `strings.Compare` is used in Prometheus code (that should be why I started using it). Would it make sense if I submit a PR to stop Prometheus from using it also?

File: vendor/github.com/prometheus/prometheus/model/labels/labels_stringlabels.go [hunk 4]
```diff
@@ -412,7 +413,7 @@ func yoloBytes(s string) (b []byte) {
 // New returns a sorted Labels from the given labels.
 // The caller has to guarantee that all label names are unique.
 func New(ls ...Label) Labels {
-	slices.SortFunc(ls, func(a, b Label) bool { return a.Name < b.Name })
+	slices.SortFunc(ls, func(a, b Label) int { return strings.Compare(a.Name, b.Name) })
```
Comment (pracucci) [inline vendor/github.com/prometheus/prometheus/model/labels/labels_stringlabels.go:416]: I would recommend to consider replacing `strings.Compare()` in this file too, and run a benchmark to see the actual perf impact.

**Comment by pracucci** [comment]
Comments on vendored Prometheus are not blockers to merge, but please consider it.

File: pkg/cardinality/request.go (hunk 1 above)
Comment (pracucci) [reply pkg/cardinality/request.go:187]: I just submitted a couple of additional comments about usage in vendored Prometheus. I think it's not a blocker to merge this PR, but I would suggest to open an upstream PR to change them (with benchmarks to justify it). If benchmark doesn't show difference, then there's no justification.

File: pkg/frontend/querymiddleware/labels_query_cache.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (aknuds1) [reply pkg/frontend/querymiddleware/labels_query_cache.go:172]: Done.

File: pkg/cardinality/request.go (hunk 1 above)
Comment (aknuds1) [reply pkg/cardinality/request.go]: Fixed.

**Comment by aknuds1** [comment]
> Please note that I'm replacing github.com/charleskorn/goautoneg with a Grafana Labs fork github.com/grafana/goautoneg, based off of the same commit. I needed to make a fix to goautoneg so it would compile with the new revision of golang.org/x/exp. I think it's better going forward to use a Grafana Labs fork rather than a personal one?

cc @charleskorn

**Pull request merged**